File: 2022/15/java/Map.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class Map {
	private final HashMap<Point, Sensor> sensors;
	private final List<Point> beacons;
	private Point pointer;

	public Map () {
		this.sensors = new HashMap<>();
		this.beacons = new ArrayList<>();
		this.pointer = new Point(0, 0);
	}

	public void addSensor(int x, int y, int bx, int by) {
		sensors.put(new Point(x, y), new Sensor(x, y, bx, by));
		beacons.add(new Point(bx, by));
	}

	public List<Sensor> getSensors() {
		return new ArrayList<Sensor>(sensors.values());
	}

	public boolean isInSensorRange(int x, int y) {
		for (Sensor s: sensors.values())
			if (s.isInRange(x, y))
				return true;
		return false;
	}

	public boolean isEmpty(int x, int y) {
		pointer.x = x;
		pointer.y = y;
		return !sensors.keySet().contains(pointer) && !beacons.contains(pointer);
	}
}

File: 2022/15/rust/src/map.rs
use crate::sensor::Sensor;
use std::collections::HashMap;
use std::fmt::{ Display, Formatter };

pub type Point = (isize, isize);

pub struct Map {
    pub sensors: HashMap<Point, Sensor>,
    pub beacons: Vec<Point>
}

impl Map {
    pub fn new() -> Self {
        let sensors = HashMap::new();
        let beacons = Vec::new();
        Self { sensors, beacons }
    }

    pub fn add(&mut self, sx: isize, sy: isize, bx: isize, by: isize) {
        self.sensors.insert((sx, sy), Sensor::new(sx, sy, bx, by));
        self.beacons.push((bx, by));
    }

    pub fn is_in_sensor_range(&self, x: isize, y: isize) -> bool {
        for s in self.sensors.values() {
            if s.is_in_range(x, y) {
                return true;
            }
        }
        false
    }

    pub fn is_empty(&self, x: isize, y: isize) -> bool {
        let p = &(x, y);
        !self.sensors.contains_key(p) && !self.beacons.contains(p)
    }
}

impl Display for Map {
    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
        for s in self.sensors.values() {
            write!(f, "{}\n", s)?;
        }
        Ok(())
    }
}
